### PR TITLE
Removing contacts from typeahead

### PR DESCRIPTION
**Pull request by martinraison**
@2is10 adds a button in the typeahead to hide an email contact. We may also want to trigger a new contact search when a contact is removed (to fill the typeahead again) - will check with product

### CODE REVIEW

File: kifi-backend/modules/shoebox/conf/shoebox.routes [hunk 1]
```diff
@@ -99,7 +99,7 @@ POST    /ext/pref/email/:kind/:send @com.keepit.controllers.ext.ExtPreferenceCon
 GET     /ext/user/friends           @com.keepit.controllers.ext.ExtUserController.getFriends()
 GET     /ext/contacts               @com.keepit.controllers.ext.ExtNonUserSearchController.findPeopleToMessage(q: String, n: Int ?= 4)
 GET     /ext/nonusers               @com.keepit.controllers.ext.ExtNonUserSearchController.findPeopleToInvite(q: String, n: Int ?= 4)
-POST    /ext/hideEmailFromUser      @com.keepit.controllers.ext.ExtNonUserSearchController.hideEmailFromUser(q: String)
+POST    /ext/hideEmailFromUser      @com.keepit.controllers.ext.ExtNonUserSearchController.hideEmailFromUser()
```
Comment (2is10) [inline kifi-backend/modules/shoebox/conf/shoebox.routes]: I think the `FromUser` is implied. How about `/ext/contacts/hide`? Going for restful, hierarchical routes.

File: packrat/scripts/lib/jquery-tokeninput.js [hunk 2]
```diff
@@ -628,6 +639,17 @@
...
+        .on('mousedown', '.kifi-dropdown-item-x', function (e) {
+          if (e.which === 1) {
+            return false;
+          }
+        })
+        .on('click', '.kifi-dropdown-item-x', function (e) {
+          if (e.which === 1) {
+            handleItemRemoved(this.parentNode);
+            return false;
+          }
```
Comment (2is10) [inline packrat/scripts/lib/jquery-tokeninput.js:684]: This is a general-purpose jQuery plugin. We should avoid mixing in concerns specific to our friend suggestion use case. It’s a little odd to see handlers being added here for events on DOM that’s not created here.

File: packrat/styles/keeper/compose.css [hunk 3]
```diff
@@ -379,6 +379,45 @@ a.kifi-compose-submit.kifi-active {
...
+  height: 18px;
+  width: 18px;
+  border-radius: 50%;
+  display: none;
+}
+.kifi-ti-dropdown-item-selected .kifi-dropdown-item-x {
+  display: block;
+}
+.kifi-dropdown-item-x:before,
+.kifi-dropdown-item-x:after {
```
Comment (2is10) [inline packrat/styles/keeper/compose.css]: `::before` and `::after` for pseudo-elements (no need for old IE compatibility here)

**Comment by 2is10** [comment]
nice. just a few minor comments. yes, I think we should refill with a new suggestion, if possible.

**Comment by martinraison** [comment]
I've done some refactoring to remove kifi-specific code from jquery-tokeninput as much as possible. It's not perfect, but I think it's better than before.

**Comment by martinraison** [comment]
Deleting a contact now also triggers a refresh of the results

File: packrat/scripts/lib/jquery-tokeninput.js [hunk 4]
```diff
@@ -657,6 +673,10 @@
 
       var query = $tokenInput.val().trim();
       if (query.length) {
+        if (oldResultsInvalid) {
+          $dropdown.find('ul').empty();
+          oldResultsInvalid = false;
+        }
```
Comment (2is10) [inline packrat/scripts/lib/jquery-tokeninput.js]: I’m curious what goes wrong if this new code is not added.

**Comment by 2is10** [comment]
Works for me.

As we discussed, would be nice to make removing a result a first-class capability of the tokeninput library.

**Comment by martinraison** [comment]
I've changed most of the logic and tried to do what we talked about yesterday. jquery-tokeninput can now be given an `onRemove` callback, to which a `replaceWith` callback is provided. The old result starts fading out as soon as the x button is clicked. If a new result is found, the new result fades in, otherwise the empty entry is closed by sliding the bottom part of the typeahead.

File: packrat/scripts/lib/jquery-tokeninput.js [hunk 5]
```diff
@@ -367,6 +378,14 @@
       return tokens.slice();
     };
 
+    this.getCurrentQuery = getCurrentQuery;
+
+    this.getItems = function () {
+      return $dropdown.find('.' + classes.dropdownItemToken).map(function (_, htmlItem) {
+        return $.data(htmlItem, 'tokenInput');
+      }).toArray();
+    }
```
Comment (2is10) [inline packrat/scripts/lib/jquery-tokeninput.js]: end assignment with semicolon

File: packrat/scripts/lib/jquery-tokeninput.js [hunk 6]
```diff
@@ -628,7 +647,35 @@
...
+
+            if ($.isFunction(settings.onRemove)) {
+              var $itemWaiting = $('<li/>')
+                .addClass(classes.dropdownItem + ' ' + classes.dropdownItemWaiting)
+                .css('display', 'none');
+              $item.after($itemWaiting);
+              var animationPromise = $item.fadeOut(200).promise().then(function () {
+                $item.remove();
+                $itemWaiting.css('display', 'block');
+                return $itemWaiting.outerHeight();
```
Comment (2is10) [inline packrat/scripts/lib/jquery-tokeninput.js:667]: what does this line do? forcing layout? why is that necessary here? might be worth commenting.

File: packrat/scripts/lib/jquery-tokeninput.js [hunk 7]
```diff
@@ -697,9 +747,36 @@
       }
     }
 
+    function replaceWith(animationPromise, item) {
+      cache.flush();
+      var $itemWaiting = this;
+      animationPromise.then(function (height) {
+        if (item) {
+            var $newEl = $(formatDropdownItem(item)).fadeIn().css('display', 'block');
+            $itemWaiting.after($newEl).remove();
```
Comment (2is10) [inline packrat/scripts/lib/jquery-tokeninput.js]: indentation off

Comment (2is10) [inline packrat/scripts/lib/jquery-tokeninput.js]: I think these two lines could be one using `.replaceWith` or `.replaceAll`:

``` js
  $(formatDropdownItem(item)).fadeIn().css('display', 'block').replaceAll($itemWaiting);
```

Is the `.css('display', 'block')` really needed? I’m pretty sure `.fadeIn()` sets it if necessary, but I don’t see how it would be here unless the stylesheet hides dropdown items by default.

File: packrat/scripts/lib/jquery-tokeninput.js (hunk 6 above)
Comment (2is10) [inline packrat/scripts/lib/jquery-tokeninput.js:667]: I see. Or maybe it would suffice to rename this to `heightPromise` or `heightAfterAnimationPromise` or something.

File: packrat/styles/keeper/compose.css [hunk 8]
```diff
@@ -343,16 +343,21 @@ a.kifi-compose-submit.kifi-active {
   margin: 0 !important;
   padding: 2px 6px 2px 42px !important;
   background: -9999px no-repeat;
-  height: 28px;
+  height: 34px;
   line-height: 28px;
+  box-sizing: border-box;
```
Comment (2is10) [inline packrat/styles/keeper/compose.css:348]: best to add `-moz-box-sizing: border-box;` too for a few more months, at least

File: packrat/styles/keeper/compose.css [hunk 9]
```diff
@@ -343,16 +343,21 @@ a.kifi-compose-submit.kifi-active {
...
 }
 .kifi-ti-dropdown-item-selected {
   background-color: #6fa6ef;
   color: #fff;
 }
+.kifi-ti-dropdown-token,
+.kifi-ti-dropdown-email {
+  border-top: 1px solid transparent;
+}
```
Comment (2is10) [inline packrat/styles/keeper/compose.css]: do you remember what this is for?

File: packrat/scripts/lib/jquery-tokeninput.js [hunk 10]
```diff
@@ -31,7 +31,10 @@
     // Callbacks
     onSelect: null,
     onAdd: null,
-    onDelete: null
+    onDelete: null,
+    onRemove: function (_, replaceWith) {
```
Comment (2is10) [inline packrat/scripts/lib/jquery-tokeninput.js]: how about we just make this `null` by default?

File: packrat/scripts/lib/jquery-tokeninput.js [hunk 11]
```diff
@@ -628,7 +647,35 @@
...
+          if (e.which === 1) {
+            return false;
+          }
+        })
+        .on('click', '.' + classes.dropdownItemX, function (e) {
+          if (e.which === 1) {
+            var $item = $(this).closest('.' + classes.dropdownItemToken);
+            var item = $.data($item[0], 'tokenInput');
+
+            if ($.isFunction(settings.onRemove)) {
```
Comment (2is10) [inline packrat/scripts/lib/jquery-tokeninput.js:659]: it looks like we don’t currently do anything at all if this function isn’t set. we should probably do something sane in that case, like just animate the element away immediately.

**Comment by 2is10** [comment]
Sweet! Feel free to merge whenever you like. I left some minor comments and suggestions, but they aren’t critical.

File: packrat/scripts/lib/jquery-tokeninput.js (hunk 7 above)
Comment (martinraison) [inline packrat/scripts/lib/jquery-tokeninput.js]: It looks like `fadeIn` sets `display` to `list-item` if the element is a `<li/>`. Not sure why, and I didn't find any mention of it in the docs. http://stackoverflow.com/questions/16885683/jquery-force-fadein-to-use-display-block-for-list-items

File: packrat/scripts/lib/jquery-tokeninput.js (hunk 5 above)
Comment (martinraison) [inline packrat/scripts/lib/jquery-tokeninput.js]: I'm spoiled by js-hint :)

File: packrat/styles/keeper/compose.css (hunk 8 above)
Comment (martinraison) [inline packrat/styles/keeper/compose.css:348]: I'm spoiled by stylus :)

File: packrat/scripts/lib/jquery-tokeninput.js (hunk 7 above)
Comment (2is10) [inline packrat/scripts/lib/jquery-tokeninput.js]: Oh, yes. I knew `.show()` did that, but forgot that other variants also did. It tries to be smart and set `display` to the element’s default  value.

File: packrat/styles/keeper/compose.css (hunk 9 above)
Comment (martinraison) [inline packrat/styles/keeper/compose.css]: I added it to make sure the `kifi-ti-dropdown-item-waiting` element would always be the same size as the element it's replacing, but the intent is not clear at all. There is probably a better way.

File: packrat/scripts/lib/jquery-tokeninput.js (hunk 10 above)
Comment (martinraison) [inline packrat/scripts/lib/jquery-tokeninput.js]: agreed

File: packrat/scripts/lib/jquery-tokeninput.js (hunk 11 above)
Comment (martinraison) [inline packrat/scripts/lib/jquery-tokeninput.js:659]: good idea
